src/core/lib/iomgr/tcp_generic_options.h

```diff
@@ -0,0 +1,117 @@
+
+#include "src/core/lib/channel/channel_args.h"
+#include "src/core/lib/iomgr/socket_mutator.h"
+#include "src/core/lib/resource_quota/api.h"
+
+typedef struct grpc_tcp_generic_options {
+  absl::flat_hash_map<absl::string_view, int> int_options;
+  grpc_core::RefCountedPtr<grpc_core::ResourceQuota> resource_quota = nullptr;
+  struct grpc_socket_mutator* socket_mutator = nullptr;
+  grpc_tcp_generic_options()
+      : resource_quota(nullptr), socket_mutator(nullptr) {}
+  // Move ctor
+  grpc_tcp_generic_options(struct grpc_tcp_generic_options&& other) noexcept {
+    socket_mutator = other.socket_mutator;
+    other.socket_mutator = nullptr;
+    resource_quota = other.resource_quota;
+    other.resource_quota.reset(nullptr);
+    int_options = other.int_options;
+  }
+  // Move assignment
+  grpc_tcp_generic_options& operator=(
+      struct grpc_tcp_generic_options&& other) noexcept {
+    if (socket_mutator != nullptr) {
+      grpc_socket_mutator_unref(socket_mutator);
+    }
+    if (resource_quota != nullptr) {
+      resource_quota.reset(nullptr);
+    }
+    socket_mutator = other.socket_mutator;
+    other.socket_mutator = nullptr;
+    resource_quota = other.resource_quota;
+    other.resource_quota.reset(nullptr);
+    int_options = other.int_options;
+    return *this;
+  }
+  // Copy ctor
+  grpc_tcp_generic_options(const struct grpc_tcp_generic_options& other) {
+    if (other.socket_mutator != nullptr) {
+      socket_mutator = grpc_socket_mutator_ref(other.socket_mutator);
+    }
+    resource_quota = other.resource_quota;
+    int_options = other.int_options;
+  }
+  // Copy assignment
+  grpc_tcp_generic_options& operator=(
+      const struct grpc_tcp_generic_options& other) {
+    if (&other == this) {
+      return *this;
+    }
+    if (other.socket_mutator != nullptr) {
+      socket_mutator = grpc_socket_mutator_ref(other.socket_mutator);
+    }
+    resource_quota = other.resource_quota;
+    int_options = other.int_options;
+    return *this;
+  }
+  // Destructor.
+  ~grpc_tcp_generic_options() {
+    if (socket_mutator != nullptr) {
+      grpc_socket_mutator_unref(socket_mutator);
+      socket_mutator = nullptr;
+    }
+    resource_quota.reset(nullptr);
+  }
+} grpc_tcp_generic_options;
+
+// Initialize the tcp generic options struct members. This method can be used
+// to initialize dynamically allocated struct objects.
+void grpc_tcp_generic_options_init(grpc_tcp_generic_options* options);
+
+// Unref members of the tcp generic options struct.
+void grpc_tcp_generic_options_destroy(grpc_tcp_generic_options* options);
+
+// Given an EndpointConfig object, returns a grpc_tcp_generic_options struct
+// constructed from it
+grpc_tcp_generic_options TcpOptionsFromEndpointConfig(
+    const grpc_event_engine::experimental::EndpointConfig& config);
+
+// Given an grpc_core::ChannelArgs object, returns a grpc_tcp_generic_options
+// struct constructed from it
+grpc_tcp_generic_options TcpOptionsFromChannelArgs(
```

Comment:
Why do we need to construct this from channel args?  Won't it always be constructed from an `EndpointConfig`?